Verify that tests in Jenkins match those in e2e.sh

spxtr opened the pull request:
This is a temporary solution in lieu of a more robust infrastructure. The test is up and running on Jenkins as kubernetes-verify-jenkins-jobs. It fails as long as there are tests in Jenkins that aren't in e2e.sh and vice versa, excluding specified exceptions and non-kubernetes tests.

@ihmccreery

ihmccreery commented on hack/jenkins/verify-jenkins-jobs.sh:
```diff
@@ -0,0 +1,88 @@
+'`
+if [ $? -ne 0 ]
+then
+  echo Could not access or read Jenkins JSON API
```
Quote.

ihmccreery commented on hack/jenkins/verify-jenkins-jobs.sh:
```diff
@@ -0,0 +1,88 @@
+  print job["name"]
+sys.exit(0)
+'`
+if [ $? -ne 0 ]
```
We tend to prefer

```console
if [[ $? -ne 0 ]]; then
```

ihmccreery commented on hack/jenkins/verify-jenkins-jobs.sh:
```diff
@@ -0,0 +1,88 @@
+  exit 1
+fi
+
+# Searches $2 for each element of $1
```
Have a look at the style guide for function comments.  Should include assumed and set global variables, maybe other things too.

ihmccreery commented on hack/jenkins/verify-jenkins-jobs.sh:
```diff
@@ -0,0 +1,88 @@
+fi
+
+# Searches $2 for each element of $1
+function search_build {
```
`function search_build() {`

lavalamp commented on hack/jenkins/verify-jenkins-jobs.sh:
```diff
@@ -0,0 +1,86 @@
+e2e_builds=$(grep "^  kubernetes-[\.|a-z|0-9|\-]*)$" "${E2E}" | tr -d " )")
+
+# Use Jenkins' JSON API to find test project names.
+jenkins_builds=$(curl -s "${JENKINS}/api/json" | python -c '
```
Can we write python programs in .py files instead of in .sh files? :)

spxtr replied:
Ok, I moved it into list-jenkins-jobs.py. Thanks.

ihmccreery commented:
@ixdy Do we need a larger conversation about taking on Python as a dependency for our test infrastructure?  I don't think it's a bad idea, (it will make a lot of stuff we're planning a lot easier,) but should be done intentionally.  I wouldn't be surprise me if there's been discussion of this in the past.

ihmccreery commented:
cc @jlowdermilk @fejta see above comment.

ixdy commented:
We already have some Python scripts in the tree, so I'm not too worried about adding another.

That said, given the small amount this Python script does, I'd suggest instead either using [jq](https://stedolan.github.io/jq/), or, as I just discovered in searching the tree for Python scripts, `build/json-extractor.py`.

ixdy commented on hack/jenkins/verify-jenkins-jobs.sh:
```diff
@@ -0,0 +1,77 @@
+#!/bin/bash
```
it'd be a good idea to set errexit, nounset, and pipefail at least. (xtrace can sometimes be helpful too, though not strictly necessary)

ixdy commented on hack/jenkins/verify-jenkins-jobs.sh:
```diff
@@ -0,0 +1,77 @@
+e2e_builds=$(grep "^  kubernetes-[\.|a-z|0-9|\-]*)$" "${E2E}" | tr -d " )")
+
+# Use Jenkins' JSON API to find test project names.
+jenkins_builds=$(curl -s "${JENKINS}/api/json" | python list-jenkins-jobs.py)
```
much of this will probably need to be reworked with errexit.

ixdy commented on hack/jenkins/verify-jenkins-jobs.sh:
```diff
@@ -0,0 +1,77 @@
+# Use Jenkins' JSON API to find test project names.
+jenkins_builds=$(curl -s "${JENKINS}/api/json" | python list-jenkins-jobs.py)
+if [[ $? -ne 0 ]]; then
+  echo "Could not access or read Jenkins JSON API"
```
echo errors to stderr: echo "..." &>2

ixdy commented on hack/jenkins/verify-jenkins-jobs.sh:
```diff
@@ -0,0 +1,77 @@
+# print them, labelled by $3.
+# Does not consider builds in EXCEPTIONS or non-kubernetes builds.
+# Returns failure through the FAIL variable.
+function search_build {
```
need parens after the function name, per style guide.

also, it'd help readability to create local variable aliases for $1, $2, and $3.

ixdy commented on hack/jenkins/verify-jenkins-jobs.sh:
```diff
@@ -0,0 +1,77 @@
+function search_build {
+  FAIL=0
+  for build in $1; do
+    grep "^kubernetes-" <(echo "${build}") > /dev/null
```
nit: I think you can collapse each of these with

```shell
if ! grep "^..." >/dev/null; then
  continue
fi
```

ixdy replied:
though actually you can probably simplify everything with a bit of pipelining (not fully tested, but the basic jist):

first turn `EXCEPTIONS` into an array:
```shell
EXCEPTIONS=(
  "kubernetes-build"
  "kubernetes-build-1.1"
  ... etc ...
  )

function join_regex() {
  local IFS="|"
  echo "$*"
}
```

and then
```shell
builds_to_consider=grep "^kubernetes-" <(echo "${build}") | grep -v "$(join_regex ${EXCEPTIONS[@]:+${EXCEPTIONS[@]}})"
```

You might be able to even fold the last grep in, though I haven't though about that enough...

spxtr commented:
I'm not a fan of adding a dependency on jq, and `build/json-extractor.py` would have to be modified to fit this use case. Is it worth it to do that when it's for such a simple and (hopefully) temporary purpose?

ixdy commented:
An even simpler option than jq or the Python script: use URL `/api/json?tree=jobs[name]&pretty=true` and grep.

jlowdermilk commented:
Re python in test scripts, my only caution is that some python libraries they not be installed on jenkins vms by default. e.g. pyyaml is a common and useful one, and not installed on vms by default. But in general I'm in favor of python over more bash.

ihmccreery commented:
:+1: To more Python less Bash.  We could maintain a K8s-test docker image with all necessary stuff installed on it, and run stuff inside there?  Is that overkill?

spxtr commented:
Thanks for the comments, @ixdy. I think I made the appropriate changes.

ixdy commented on hack/jenkins/verify-jenkins-jobs.sh:
```diff
@@ -0,0 +1,77 @@
+jenkins_builds=$(curl -sg "${JENKINS}/api/json?tree=jobs[name]&pretty=true" \
+  | grep -Po '(?<="name" : ")[^"]*')
+search_build "${e2e_builds}" "${jenkins_builds}" "Jenkins"
+fail_jenkins=${FAIL}
```
small suggestion: maybe rather than relying on the FAIL variable, make search_build return this value?

spxtr replied:
Won't that make errexit trigger if the first test fails? I guess I can return it as a string.

ixdy replied:
good point. you can do `search_build ... && fail_jenkins=0 || fail_jenkins=1`, though that is pretty ugly too.

spxtr replied:
`exit $(search_build ... || search_build ...)` works :)

spxtr commented:
@k8s-bot e2e test this

ixdy commented:
LGTM. Please squash.

spxtr committed:
Verify that tests in Jenkins match those in e2e.sh.

This is a temporary solution in lieu of a more robust infrastructure.
The test is up and running on Jenkins as kubernetes-verify-jenkins-jobs.
It fails as long as there are tests in Jenkins that aren't in e2e.sh and
vice versa, excluding specified exceptions and non-kubernetes tests.

krousey commented:
Manually merging while merge queue is broken. This doesn't seem like it will break anything and only improve things.
